src/lfo.rs

```diff
@@ -0,0 +1,115 @@
+// implements a wavetable LFO
+
+
+use std::f32::consts::PI;
+use crate::ring_buffer::RingBuffer;
+
+/// LFO is a struct that contains a wavetable and a phase, frequency, and amplitude.
+pub struct LFO {
+    sample_rate: f32,
+    wavetable: RingBuffer<f32>,
+    phase: f32,
+    frequency: f32,
+    amplitude: f32,
+
+}
+/// implements functions for LFO struct
+impl LFO {
+    /// creates a new LFO with a given sample rate, frequency, and amplitude, only does so for one period
+    pub fn new(sample_rate: f32, frequency: f32, amplitude: f32) -> LFO {
+        let wavetable_size = (sample_rate / frequency) as usize;
+        let mut wavetable = RingBuffer::new(wavetable_size);
+        for i in 0..wavetable_size {
+            let value = amplitude * (2.0 * PI * i as f32 / wavetable_size as f32).sin();
+            wavetable.push(value);
+        }
+        LFO {
+            sample_rate,
+            wavetable,
+            phase: 0.0,
+            frequency,
+            amplitude: 1.0,
+        }
+    }
+
```

Comment:
With a wavetable oscillator, you shouldn't have to make a new waveform every time you change the frequency. You can just initialize a 1Hz sine wave in your buffer and use pop_frac and read at iterations of the frequency. This way you never have to change the buffer. You can just interpolate between the values for different frequencies.

Reply:
In that sense, would we introduce more errors into the system? Since we are interpolating a sine wave in a linear way